sorting/radix_sort.c

```diff
@@ -1,71 +1,69 @@
-    int i, j, k, remainder, NOP = 0, divisor = 1, large, pass;
-
-    large = largest(a, n);
-    printf("The large element %d\n", large);
-    while (large > 0)
-    {
-        NOP++;
-        large /= 10;
-    }
-
-    for (pass = 0; pass < NOP; pass++)
-    {
-        for (i = 0; i < 10; i++)
-        {
-            bucket_count[i] = 0;
-        }
-        for (i = 0; i < n; i++)
-        {
-            remainder = (a[i] / divisor) % 10;
-            bucket[remainder][bucket_count[remainder]] = a[i];
-            bucket_count[remainder] += 1;
-        }
-
-        i = 0;
-        for (k = 0; k < 10; k++)
-        {
-            for (j = 0; j < bucket_count[k]; j++)
-            {
-                a[i] = bucket[k][j];
-                i++;
-            }
-        }
-        divisor *= 10;
-
-        for (i = 0; i < n; i++) printf("%d  ", a[i]);
-        printf("\n");
-    }
-}
-
-int main()
-{
-    int i, n, a[10];
-    printf("Enter the number of elements :: ");
-    scanf("%d", &n);
-    printf("Enter the elements :: ");
-    for (i = 0; i < n; i++)
-    {
-        scanf("%d", &a[i]);
-    }
-    RadixSort(a, n);
-    printf("The sorted elements are ::  ");
-    for (i = 0; i < n; i++) printf("%d  ", a[i]);
-    printf("\n");
-    return 0;
-}
+#include <stdio.h>
+
+// To find the largest number in the array
+int findLargestNum(int data[], int length) {
+    int largest = data[0];
+    for (int i = 1; i < length; i++)
+        if (data[i] > largest)
+            largest = data[i];
+    return largest;
+}
+
+// To perform counting sort on the array based on the specified digit
+void performCountSort(int data[], int length, int place) {
+    int output[length]; 
+    int i, freq[10] = {0};
+    for (i = 0; i < length; i++)
+        freq[(data[i] / place) % 10]++;
+    for (i = 1; i < 10; i++)
+        freq[i] += freq[i - 1];
+    for (i = length - 1; i >= 0; i--) {
+        output[freq[(data[i] / place) % 10] - 1] = data[i];
+        freq[(data[i] / place) % 10]--;
+    }
+    for (i = 0; i < length; i++)
+        data[i] = output[i];
+}
+
+// Main function for Radix Sort
+void performRadixSort(int data[], int length) {
+    int largestNum = findLargestNum(data, length);
+    for (int place = 1; largestNum / place > 0; place *= 10)
+        performCountSort(data, length, place);
+}
+void displayArray(int data[], int length) {
+    for (int i = 0; i < length; i++)
+        printf("%d ", data[i]);
+    printf("\n");
+}
+int main() {
+    int array1[] = {121, 432, 564, 23, 1, 45, 788};
+    int length1 = sizeof(array1) / sizeof(array1[0]);
+    int array2[] = {95, 90, 85, 80, 75, 70};
+    int length2 = sizeof(array2) / sizeof(array2[0]);
+    int array3[] = {333, 287, 911, 462, 786};
+    int length3 = sizeof(array3) / sizeof(array3[0]);
+    int array4[] = {5, 3, 8, 1, 2, 9, 4, 7, 6};
+    int length4 = sizeof(array4) / sizeof(array4[0]);
+    int array5[] = {12, 34, 12, 45, 67, 45, 34};
+    int length5 = sizeof(array5) / sizeof(array5[0]);
+

```

Comment:
if you are going for speed with the test of this algorithm you could just do 
```c
float size_of_int = 1/sizeof(int);
// then multiply this value instead of divding over and over again
// remember multiplication is always faster than division
int length1 = sizeof(array1) * size_of_int;
int length2 = sizeof(array2) * size_of_int;
int length3 = sizeof(array3) * size_of_int;
int length4 = sizeof(array4) * size_of_int;
int length5 = sizeof(array5) * size_of_int;